interop-cats/jvm/src/main/scala/zio/interop/ZioAsync.scala

```diff
@@ -0,0 +1,59 @@
+package zio.interop
+
+import cats.effect.kernel.Unique
+import cats.effect.{ Async, Cont, Sync }
+import zio.blocking.{ effectBlocking, effectBlockingInterrupt, Blocking }
+import zio.clock.Clock
+import zio.internal.Executor
+import zio.{ RIO, ZIO }
+
+import scala.concurrent.{ ExecutionContext, Future }
+
+private class ZioAsync[R <: Clock with Blocking] extends ZioTemporal[R, Throwable] with Async[RIO[R, *]] {
+
+  override final def evalOn[A](fa: F[A], ec: ExecutionContext): F[A] =
+    ZIO.effectSuspendTotalWith { (platform, _) =>
+      fa.lock(Executor.fromExecutionContext(platform.executor.yieldOpCount)(ec))
+    }
+
+  override final val executionContext: F[ExecutionContext] =
+    ZIO.effectSuspendTotalWith { (platform, _) =>
+      ZIO.succeedNow(platform.executor.asEC)
+    }
+
+  override final val unique: F[Unique.Token] =
+    ZIO.effectTotal(new Unique.Token)
+
+  override final def cont[K, Q](body: Cont[F, K, Q]): F[Q] =
+    Async.defaultCont(body)(this)
+
+  override final def suspend[A](hint: Sync.Type)(thunk: => A): F[A] = hint match {
+    case Sync.Type.Delay                                           => ZIO.effect(thunk)
+    case Sync.Type.Blocking                                        => effectBlocking(thunk)
+    case Sync.Type.InterruptibleOnce | Sync.Type.InterruptibleMany => effectBlockingInterrupt(thunk)
+  }
+
+  override final def delay[A](thunk: => A): F[A] =
+    ZIO.effect(thunk)
+
+  override final def defer[A](thunk: => F[A]): F[A] =
+    ZIO.effectSuspend(thunk)
+
+  override final def blocking[A](thunk: => A): F[A] =
+    effectBlocking(thunk)
+
+  override final def interruptible[A](many: Boolean)(thunk: => A): F[A] =
+    effectBlockingInterrupt(thunk)
+
+  override final def async[A](k: (Either[Throwable, A] => Unit) => F[Option[F[Unit]]]): F[A] =
```

Comment:
I think we need a slightly more completed state machine here to honor the guarantee expected by Cats Effect that if the `F` returned by the register function fails the return value of the callback will not be observed. Something like:

```scala
    Promise.make[Nothing, Unit].flatMap { promise =>
      ZIO.effectAsyncM { register =>
        k(either => register(promise.await *> ZIO.fromEither(either))) *> promise.succeed(())
      }
    }
```

Reply:
Awesome 👍